main/prod/python/tradingBot.py

```diff
@@ -0,0 +1,164 @@
+        #calculating de date for the first candle of the dataset
+        start_date = management.calc_start_date(strategy)
+
+        #getting intraday candle dataset from binance
+        intraday_data = CandleData(pair, strategy.intraday_interval, start_date)
+        intraday_data.populate_data(round(time.time()*1000))
+        
+        #adding strategy indicators to intraday dataset
+        intraday_dataset = Dataset(intraday_data.candle_df, strategy)
+        intraday_dataset.add_indicators_to_candle_dataset("intraday")
+
+        #getting trend candle dataset from binance
+        trend_data = CandleData(pair, strategy.trend_interval, start_date)
+        trend_data.populate_data(round(time.time()*1000))
+
+        #adding strategy indicators to trend dataset
+        trend_dataset = Dataset(trend_data.candle_df, strategy)
+        trend_indicators_list = trend_dataset.add_indicators_to_candle_dataset("trend")
+
+        #merging intraday and trend datasets in one final dataset
+        return intraday_dataset.merge_dataframes(trend_dataset.dataset, *trend_indicators_list)
+
+
+    def handle_new_trades(self):
+        # Check if the bot is active
+        if not self.setup.opperation_active:
+            return
+
+        self.active_pairs = self.db.get_active_pairs()
+
+        #as the get_open_trade_pairs returns 2 values, just get the first one(pair)
+        opened_trade_pairs = [pair[0] for pair in self.db.get_open_trade_pairs()]
+
+        # TODO: get pairs with opened alerts
+
+        # Select eligible pairs.
+        # Pairs that are active, do not have any opened position and do not have any opened alert.
+        self.pairs = [pair for pair in self.active_pairs if pair not in opened_trade_pairs]
+
+        for pair in self.pairs:
+            for strategy in self.strategies:
+                # Check if the strategy has been executed recently
+                if not self.should_run_strategy(pair, strategy):
+                    # Jump to the next strategy for this pair.
+                    continue
+
+                # Updates the last run time
+                self.last_executions[(pair, strategy)] = datetime.now()
+                final_dataset = self.create_combined_dataset(pair, strategy)
+
+                #logging for debugging
+                logger.info(f"TRYING TO OPEN POSITION")
+                logger.debug(f"Pair: {pair} - datetime: {datetime.now()} - final_dataset:\n{final_dataset}")
+
+                manager = StrategyManager(
+                    pair,
+                    final_dataset,
+                    self.exchange_session.e_id,
+                    self.exchange_session.e_sk,
+                    self.setup.order_value,
+                    strategy
+                )
+                manager.try_open_position()
+
+
+    def handle_opened_trades(self):
+        # Handle opened trades. Check if we is ready to sell.
+        self.opened_trade_pairs = db.get_open_trade_pairs()
```

Comment:
Do jeito que o código está, vc tá indo no banco pra pegar os trades abertos. 
E depois dentro do `for loop`, indo no banco novamente (na mesma tabela) pra pegar o `strategy_id`.

Aí dentro do `for loop`, vc consegue iterar a lista de trades. Repare que o `strategy_id` já vem de brinde!

Reply:
Outra coisa, dentro do `for loop` tem outra chamada no banco pra pegar o `strategy_name` (method: `db.get_strategy_name`). 
Não é uma boa prática fazer chamadas ao banco dentro de loops. O ideal seria fazer a chamada do `get_strategy_name` fora do loop, salvar os dados necessários num dicionário. 
E aí dentro do loop, quando precisar pegar o strategy_name, acessa o dicionario passando o `trade.strategy_id`

Reply:
Refatorando, ficaria algo como:

```
def handle_opened_trades(self):
        # Handle opened trades. Check if is ready to sell.
        self.trades = db.get_open_trade_pairs()

        # get all used strategy_ids - Using the "set" concept to remove duplicate strategy IDs.
        strategy_ids = {trade.strategy_id for trade in self.trades}
        strategy_ids_list = list(strategy_ids)  # Convertendo para lista

        # get all used stragies
        # Here I assume the db.get_strategies receives a list of ids, and return a list of strategy object (with all properties)
        self.strategies = db.get_strategies(strategy_ids_list)
        
        # Converts the list of strategies to a dictionary in the format {strategy_id: strategy_name}
        self.strategy_dict = {strategy.id: strategy.name for strategy in self.strategies}

        for trade in self.trades:
            strategy_name  = self.strategy_dict[trade.strategy_id] 
            strategy_instance = strategy_name()
```
Repara que toda chamada ao banco foi feita antes do loop. Isso deixa o código bem mais eficiente e rápido!